flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLoggingIterator.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.function.BiConsumerWithException;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+class StateChangeLoggingIterator<State, StateElement, Namespace> implements Iterator<StateElement> {
+
+    private final Iterator<StateElement> iterator;
+    private final StateChangeLogger<State, Namespace> changeLogger;
+    private final BiConsumerWithException<StateElement, DataOutputViewStreamWrapper, IOException>
+            removalWriter;
+    private final Namespace ns;
+    @Nullable private StateElement lastReturned;
+
+    private StateChangeLoggingIterator(
+            Iterator<StateElement> iterator,
+            StateChangeLogger<State, Namespace> changeLogger,
+            BiConsumerWithException<StateElement, DataOutputViewStreamWrapper, IOException>
+                    removalWriter,
+            Namespace ns) {
+        this.iterator = iterator;
+        this.changeLogger = changeLogger;
+        this.removalWriter = removalWriter;
+        this.ns = ns;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return iterator.hasNext();
+    }
+
+    @Override
+    public StateElement next() {
+        return lastReturned = iterator.next();
+    }
+
+    @Override
+    public void remove() {
+        iterator.remove();
+        try {
+            changeLogger.valueElementRemoved(out -> removalWriter.accept(lastReturned, out), ns);
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e);
+        }
+    }
+
+    @Nullable
+    public static <Namespace, State, StateElement> Iterator<StateElement> create(
+            @Nullable Iterator<StateElement> iterator,
+            StateChangeLogger<State, Namespace> changeLogger,
+            BiConsumerWithException<StateElement, DataOutputViewStreamWrapper, IOException>
+                    removalWriter,
+            Namespace ns) {
+        return iterator == null
+                ? null
```

Comment:
We should return `Collections.emptyIterator()` here and tag this method as `@Nonnull` which aligns with `InternalPriorityQueue#iterator`